### DISCUSSION TITLE
Control board and Auger rate mods

### DISCUSSION THREAD

**Discussion by jazzmonger** [jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip#15] [General]
The control board has several opto  isolator Ics that control triacs that control all the ac motors and igniter.  The triacs are 136SE6 which are rated for 4A, so at 120v they can sink 480watts.  This is good bc the new igniter I got from Amazon is 400w, and it works even better than the orginial. But, it sticks out into the burn pot, so it's not ideal for long term use as I'll have to retract it to remove the burn pot for cleaning.  I'm STILL waiting on the replacement igniter from the vendor, so its better than manually lighting the burn pot with Map Gas....

Christ, is anyone but me ever going to read all this?  My wife says I'm completely obsessed with "The Pellet Stove". Perhaps when I wake up at 5:28am to fire up HOME ASSISTANT and watch all my automations and settings take place at 5:30am . Perhaps she's right.  
Anyway, moving Onward....

![AA923C38-5124-4F0F-8CBE-9F4E1A92397B](https://user-images.githubusercontent.com/52110065/204321900-1d7f7ce5-9806-46b8-b2e7-702a8379ebf2.jpeg)

These triacs are controlled by the white opto isolator's pin 2 directly connected to the MCU's gpio outputs.  They turn ON the AC device they control  when pin2 goes from HIGH to LOW or 1/ON to OFF/0.

![6A6824FA-67FD-49F3-88D5-EDEEEBE577C6](https://user-images.githubusercontent.com/52110065/204325219-bb1ca680-4b74-469f-a674-2974f68300e5.jpeg)

igniters: (labeled "I")
The std igniter is 300w, so most common igniters up to 400w should work, drawing 3.2A from the triac, leaving a 15% margin of the 4A max triac current capacity.  These stoves have 6" x 3/8" 300w igniters held in place by 2 Phillips head screws.  Mine failed a week ago and it takes a couple weeks to get these so I ordered some extra ones to have on hand for the next failure. 

Auger (labeled "f"=feed")
ECO2 mode starts at P1 and when temp is reached throttles down to P4. In a large area like my shop, this works well.  The stove burns continuously at P4 unless it gets colder than set.temp-3 degrees then switches back to P1 until current.temp=set.temp+1. And the cycle then repeats. 

EDIT: I verified with my spiffy new Logic Analyzer that this is indeed how the stove operates.

For smaller areas, after the room reaches settemp and throttles down from P1 to P4, the room often gets overheated without the stove turning off. That's just the way ECO2 works deep in the internals of the main stove control board. 

Now that I know how to control the triac for the auger, my future plan is to cut the trace on the auger control signal and monitor and echo the signal from the MCU auger pin and inject it into the opto-isolator from a Tywe1s gpio pin.  Then when I choose a new level P5 that I will define in the Power select in ESPHome, it that will shorten the on time and lengthen the off time of the auger as long as the exhaust temp is above say 250 degrees.  P4 cruises along at about 270 under normal operation.  It will be best to do this directly in the ESP using ESPHome instead of an automation just in case home assistant crashes or becomes unavailable.

I'll leave the on time the same so pellets don't get stuck. I should be able to use 2 of the unused GPIOs in the TYWE1S for this.

Fan (labeled "w")
This fan is controlled by the triac I labeled "w" and goes from 107vac on P1 to 97v on P4 thus varying the speed of the hopper fan.I hooked a std box fan to this output to figure this out. The "S" setting in the menu controlling the various P levels is what controls this triac.

**Comment by jazzmonger** [comment]
After about 6 weeks, I finally got my logic Analyzer from China!  Sheesh, I'm an EE and it took me this long to finally get a logic analyzer.  I put it to good use yesterday and monitored all of the opto-isolators as the stove goes through its cycles.  All values are in seconds.  P5 is what I'm considering trying first.

![BE82490D-A09E-4E40-826D-EB7A918534FD_1_201_a](https://user-images.githubusercontent.com/52110065/205527250-330492d8-4eb4-4488-8c5a-b288981e0824.jpeg)

Here are the definitive auger on/off times I measured on the Logic Analyzer for each of the Power Levels P1-P4

<img width="345" alt="image" src="https://user-images.githubusercontent.com/52110065/226465354-29b80503-8b41-478e-b15a-5e88303b45a6.png">

Hopefully, I'll use this info to add a new "Ultra Low" P5 program that really slows down the burn, but allows me to use ECO2 and keep the stove burning instead of turning it off and turning it back on.

My plan is to cut the trace on the Auger opto-isolator input and use one of the GPIOs on the TYWE1S as its input to echo the auger signal info.  This will let me create the "new" P5 plan, and then I'll play w/ the value until I get the burn I want.  Time will tell!

**Comment by jazzmonger** [comment]
After a few days of intense coding, consulting with devs on how to do it and lots of trial and error I finally have the new fully configurable P5 power level with full auger echo on P1-P4 control working flawlessly. The biggest challenge was to do it all inside the tywe1s chip running ESPHome with zero reliance on home assistant. If home assistant crashes, we don't want the Auger to feed pellets non-stop to a burning pot.  Nothing good would come from that. So, this approach is employed with safety in mind. 

I created 2 number helpers as sliders that will help me fine tune the new P5 power level. For safety they are set at the upper and lower on/off times I measured in all power levels. As I understand what the burn rates actually do by watching the stove and temps and I get the lower  pburn rate  I'm after working well, I'll likely hard code these values and set them permanently in the ESPHome config file. 

![373315F1-76DA-4ED5-BDDE-4F9576B9EB85](https://user-images.githubusercontent.com/52110065/207506637-3c99e842-3088-4dae-87dc-d44bf09a2c5f.jpeg)

when the P5 switch is OFF, the Auger output to the opto isolator echos the Auger input from the MCU. It's like nothing changed.  When ON, my new power level routine will slow down the the pellet delivery and lengthen the time before new pellets are added to the burn pot.  Hopefully This will cool the stove down some so that I can keep it lit and reduce the on/off cycles that destroy the igniter.  

I chose this approach because the auger control relies on lots of variables.  Are we lighting? Are we in an error state?  Are we shutting down? Echoing the signal for P1-P4 seems like the safest approach.

This mod requires the trace on the MCU board to be cut on the pins from the MCU that controls the Auger opto isolator.  We insert a new GPIO to monitor the auger signal from the MCU and another GPIO to output control of the Auger opto isolator which turns the Auger motor on and off. Two new wires are run from the GPIOs on the tywe1s chip in the display unit to the MCU board for this mod to function.

I also added an ultrasonic hopper pellet level sensor. It works, but it's disconnected for now which is why it says unavailable.

I'll post this new well-documented ESPHome config file in the uploads section.

**Comment by Homewatts** [comment]
I love the work you are doing, and sharing. Keep at it! Your wife may think you're crazy but I don't. It's supper satisfying and fun to delve into this kind of work.
Personally, I'm looking at replacing my auger motor with a variable speed auger motor, and modifying the control board firmware to allow the control of this new motor. I want to do this because I hate the noise that the auger makes turning on and off.
Do you by chance have source code for the control board?
I have retrieved the binary (Hex) file from the Atmel Atmega16a chip and have converted it to an assembly file hoping to reconstruct a C file. Reverse engineering this into a C file will be a huge undertaking.
Any ideas?
Thanks, Charlie

**Reply by jazzmonger** [reply]
Charlie,
the last time I used assembly language was on my 8088 bas PC in college with a reset button on the front!  You're a better man than I am :)

You might be better off creating a soft start circuit for the auger motor and/or isolating it's body with rubber pads.  the noise I hear is 60hz, so it's the hum of the field windings.  The triacs that control the fans are varying their voltage so this would be easier to do on the auger motor than debugging assembly!  even lowering the field voltage might lower the noise & speed of the auger.  then use my mods to compensate for the reduced speed by increasing the ON time.  Just a thought.

**Comment by Geo-Ron** [comment]
Fyi: people ARE reading this!

**Reply by jazzmonger** [reply]
I see that!   I fired up both my stoves yesterday as it got below freezing here.  Both still working like champs.
